### PR TITLE
[#151370067] Adds ADR 10 on release and reployment process of Azure Functions

### PR DESCRIPTION
**Pull request by cloudify**

### CODE REVIEW

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+The Digital Citizenship APIs are based on the Azure Functions framework (a _serverless_ application framework). 
+
+During the initial development of the APIs, we relied on the [continuous deployment feature](https://docs.microsoft.com/en-us/azure/azure-functions/functions-continuous-deployment), configured to deploy the API code from the `master` branch of the [digital-citizenship-functions](https://github.com/teamdigitale/digital-citizenship-functions) repository.
+
+This release process served us well during the initial development phase, but it has several drawbacks:
+
+1. there's no approval process that triggers the deployment of the `master` branch into production
+1. the build phase happens on the production hosts
+1. we cannot control the actual deployment process (the build and the re-load of the code happens automatically on the production hosts)
+1. Rolling back is cumbersome as it requires revering commits on the `master` branch.
```
Comment (gunzip) [inline docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: typo: reverting

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+On tests passing successfully, the CI workflow triggers the build process that generates artifacts needed for release (e.g. packs dependencies, compiles Typescript into Javascript, etc...).
+
+The generated artifact get stored in a `build` directory (e.g. a single javascript file generated by [Azure Functions Pack](https://github.com/Azure/azure-functions-pack)).
+
+### 3 - Version bump and tags
+
+On successful build, the patch version gets incremented (e.g. in `package.json`) and it gets committed along with the generated artifacts. 
+
+After commit, two tags get created:
```
Comment (gunzip) [inline docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: git tags ?

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+* a `SNAPSHOT-x.x.x` tag using the bumped version
+* a `SNAPSHOT-LATEST` tag (this tag gets updated to the current commit)  
+
+Everything gets pushed to the upstream repository.
+
+### 4 - Promotion
+
+On the CI workflow, a promotion step awaits for approval.
```
Comment (gunzip) [inline docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: it's not clear how

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+We decide to structure the release and deployment process in the following steps:
+
+### 1 - Automated tests
+
+On commit to the `master` branch, the CI system checks out the commit and runs the automated tests.
+
+### 2 - Build of release artifacts
+
+On tests passing successfully, the CI workflow triggers the build process that generates artifacts needed for release (e.g. packs dependencies, compiles Typescript into Javascript, etc...).
```
Comment (gunzip) [inline docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: we cite a  'CI workflow' but it looks like it's not described anywhere

Comment (cloudify) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: yes, meaning we need to create on, I'll rephrase

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md (hunk 2 above)
Comment (cloudify) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: yes

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md (hunk 3 above)
Comment (cloudify) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: yes sorry, I should add links, I'm talking about [circleci workflows](https://circleci.com/docs/2.0/workflows/)

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+Everything gets pushed to the upstream repository.
+
+### 4 - Promotion
+
+On the CI workflow, a [promotion step](https://circleci.com/docs/2.0/workflows/#holding-a-workflow-for-a-manual-approval) awaits for approval.
+
+When the promotion gets triggered, two git tags get created:
+
+* a `RELEASE-x.x.x` tag using the current version
+* a `RELEASE-LATEST` tag (this tag gets updated to the current commit)
```
Comment (gunzip) [inline docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: it looks like you cannot set up continuos deployment from a tag. 
release-latest and snapshot-latest must be git branches.

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+### 5 - deployment
+
+Deployment will still happen through the [continuous deployment feature](https://docs.microsoft.com/en-us/azure/azure-functions/functions-continuous-deployment) of Azure Functions. 
+
+The difference introduced with this new model is that instead of linking the continuous deployment to the `master` branch, we will link it to a git tag that we can control independently from what gets committed to the `master` branch. Also we could use a different git tag for each environment (i.e. staging and development).
+
+#### staging
+
+The staging environment will be configured for continuous deployments from the `SNAPSHOT-LATEST` tag.
```
Comment (gunzip) [inline docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: it's not clear if the staging environment it's a function slot and which resources (ie. cosmosdb, queues) shares with the production environment.

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+#### production
+
+The production environment will be configured for continuous deployments from the `RELEASE-LATEST` tag.
```
Comment (gunzip) [inline docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: it's not clear how migrations are handled. ie. what if you have to add a cosmosdb collection or queue ?

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md (hunk 5 above)
Comment (cloudify) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: ufffff branches are a little more difficult to manage in this way, need to reset the branch to the commit, not very clean, but let's see if we can manage to make it work

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md (hunk 7 above)
Comment (cloudify) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: that would be done before the deployment, with terraform - there's a certain time interval in which the old code will run on the "new" infrastructure (or vice versa: new code should run on old infra), code should handle that - a common strategy is to use feature flags

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md (hunk 6 above)
Comment (cloudify) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: in this case is a completely different infrastructure, completely separated from the production one (as I mentioned [here](https://github.com/teamdigitale/digital-citizenship/pull/17#discussion_r146107023)) we will create two infrastructures that share the same setup, changes will be first applied to staging and then to production, they will not share anything.

File: docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md (hunk 5 above)
Comment (cloudify) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: btw somebody asked for this a [while ago](https://github.com/projectkudu/kudu/issues/1435)

Comment (gunzip) [reply docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md]: as a workaround, we can use the continous deployment integration just as a  "trigger" then plug some custom logic into deploy.cmd which will do a git checkout of the master#RELEASE **tag** and copy the files to wwwroot.

**Comment by cloudify** [comment]
This has been superseded by https://github.com/teamdigitale/digital-citizenship-functions/pull/87, closing

**Pull request closed without merging**